File: theory/graphs/disjoint_set.py
## Disjoint Set

class disjoint_set:
    def __init__(self, n):
        self.parent = [i for i in range(n)]

    def find(self, x):
        if self.parent[x] == x:
            return x
        
        ## Path compression
        self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def union(self, x, y):
        a = self.find(x)
        b = self.find(y)
        self.parent[b] = a

class Disjoint_Set:
    def __init__(self,)
        
def main():
    ds_obj = disjoint_set(5)
    ds_obj.union(1, 2)
    ds_obj.union(1, 3)
    ds_obj.union(0, 4)
    print(ds_obj.find(4))
    print(ds_obj.parent)

main()

File: algorithm-notes/graphs/disjoint_set.cpp
#include <iostream>
using namespace std;

class DisjointSet {
    vector<int> parent;
public:
    DisjointSet(int n) {
        for (int i = 0; i <= n; i++)
            parent.push_back(i);
    }
    int Find(int node) {
        if (parent[node] == node) return node;
        return parent[node] = this->Find(parent[node]);
    }
    void Union(int x, int y) {
        int a = this->Find(x);
        int b = this->Find(y);
        parent[b] = a;
    }
};

int main() {
    DisjointSet ds(5);
    ds.Union(1, 2);
    ds.Union(3, 4);
    ds.Union(0, 5);
    ds.Union(4, 0);
    cout << ds.Find(2) << "\n";
    return 0;
}